### PR TITLE
Added endpoint for company leaderboard

### PR DESCRIPTION
**Pull request by Uttkarsh-raj**
Added endpoints to get a scoreboard/ leaderboard for the companies according to the total number of issues.
Will help in implementing and tracking the leaderboard for the companies in the application.

Screenshot (116)

### CODE REVIEW

**Review by arkid15r — COMMENTED** [review]
Could you check if the following suggestions make sense in this context:

File: website/api/views.py [hunk 1]
```diff
@@ -400,19 +401,40 @@ def global_leaderboard(self, request, *args, **kwargs):
...
+        )
+
+        companies = []
+        for url in company_urls:
+            domain = Domain.objects.filter(id=url["domain"]).first()
+            company = Company.objects.filter(url=domain.url).first()
+            serializer = CompanySerializer(company)
+            companies.append({"company": serializer.data, "issue_count": url["issue_count"]})
+            if len(companies) >= 100:
+                break
```
[the remark this answers is not in the document]
Comment (Uttkarsh-raj) [reply website/api/views.py:434]: makes sense... ill try this.. i tried it with query but was not getting the results.. but ill try to optimize this..

File: website/api/views.py [hunk 2]
```diff
@@ -400,19 +401,40 @@ def global_leaderboard(self, request, *args, **kwargs):
...
-
+        elif leaderboard_type == "companies":
+            return self.company_leaderboard(request, *args, **kwargs)
         else:
             return self.global_leaderboard(request, *args, **kwargs)
 
+    def company_leaderboard(self, request, *args, **kwargs):
+        company_urls = (
+            Issue.objects.values("domain")
+            .annotate(issue_count=Count("domain"))
```
[the remark this answers is not in the document]
Comment (Uttkarsh-raj) [reply website/api/views.py:423]: Have tried this locally and it was working but i guess ill have to check it with @AtmegaBuzz ...

**Comment by AtmegaBuzz** [comment]
@Uttkarsh-raj
1. ```@action``` decorator can be only used inside modelviewset and not api view 
2. as i already told you, you can easily chain foreignkeys to fetch the values from parent 

Here Company is parent of Domain and DOmain is parent of Issue
Company -> DOmain -> Issue
so the parents above the hierarchy can access their child values using __
if 'X' has 5 level child then you have to add 5 _ eg "domain__issue__x__y__zfield"
 
```python
def company_leaderboard(self, request, *args, **kwargs):
        
        paginator = PageNumberPagination()
        companies = Company.objects.values().annotate(issue_count=Count('domain__issue')).order_by("-issue_count")
        page = paginator.paginate_queryset(companies, request)

        return paginator.get_paginated_response(page)
```

**Comment by Uttkarsh-raj** [comment]
> 1. `@action` decorator can be only used inside modelviewset and not api view
> 2. as i already told you, you can easily chain foreignkeys to fetch the values from parent
> Try this logic it is db level optimised with pagination

This do seems to work . Thanks for the insight , this is my first time with working with this orm, i think i came up with the correct sql query and maybe i was not able to use the orm optimally thanks for the review @AtmegaBuzz . Working on it.

Also do we have to show all the companies in the leaderboard or only the top 100 or so? @DonnieBLT

**Comment by AtmegaBuzz** [comment]
No worries @Uttkarsh-raj, every time you make a pr I will try to give an optimised orm query to give a new perspective of how powerful it is, so that you can learn from it. 

Api view has a get method and you can override it. Modelviewset doesn't have get instead it uses list and show. I think the api view is also fine if the current implementation works.

**Comment by Uttkarsh-raj** [comment]
> Api view has a get method and you can override it. Modelviewset doesn't have get instead it uses list and show. I think the api view is also fine if the current implementation works.

```

class APIView(View):

    @classmethod
    def as_view(cls, **initkwargs):

    @property
    def allowed_methods(self):

    @property
    def default_response_headers(self):

    def http_method_not_allowed(self, request, *args, **kwargs):

    def permission_denied(self, request, message=None, code=None):

    def throttled(self, request, wait):

    def get_authenticate_header(self, request):

    def get_parser_context(self, http_request):

    def get_renderer_context(self):

    def get_exception_handler_context(self):

    def get_view_name(self):

    def get_view_description(self, html=False):

    def get_format_suffix(self, **kwargs):

    def get_renderers(self):

    def get_parsers(self):

    def get_authenticators(self):

    def get_permissions(self):

    def get_throttles(self):

    def get_content_negotiator(self):
 
    def get_exception_handler(self):

    def perform_content_negotiation(self, request, force=False):

    def perform_authentication(self, request):

    def check_permissions(self, request):

    def check_object_permissions(self, request, obj):

    def check_throttles(self, request):

    def determine_version(self, request, *args, **kwargs):

    def initialize_request(self, request, *args, **kwargs):

    def initial(self, request, *args, **kwargs):

    def finalize_response(self, request, response, *args, **kwargs):

    def handle_exception(self, exc):

    def raise_uncaught_exception(self, exc):

    def dispatch(self, request, *args, **kwargs):

    def options(self, request, *args, **kwargs):

```
I could find only the above methods in the APIView. And i tried overriding the get api which is not getting the required results.
Thank you.

**Comment by AtmegaBuzz** [comment]
See here, api view inherits get api view, post api ... That's why you didn't see it cause it's inside the parent classes

**Comment by AtmegaBuzz** [comment]
Looks good, Thanks @Uttkarsh-raj

**Review by DonnieBLT — APPROVED** [review]

**Comment by Uttkarsh-raj** [comment]
> Looks good, good work @Uttkarsh-raj

Thank you. Really appreciate the reviews.

**Pull request merged**